daemon/swap_daemon.go

```diff
@@ -0,0 +1,165 @@
+		panic("swap factory address not specified")
+	}
+
+	ec := conf.EthereumClient
+	chainID := ec.ChainID()
+
+	// Initialize the database first, so the defer statement that closes it
+	// will get executed last.
+	sdb, err := db.NewDatabase(&chaindb.Config{
+		DataDir: path.Join(conf.EnvConf.DataDir, "db"),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if dbErr := sdb.Close(); dbErr != nil {
+			err = multierror.Append(err, fmt.Errorf("syncing database: %s", dbErr))
+		}
+	}()
+
+	sm, err := swap.NewManager(sdb)
+	if err != nil {
+		return err
+	}
+
+	hostListenIP := "0.0.0.0"
+	if conf.EnvConf.Env == common.Development {
+		hostListenIP = "127.0.0.1"
+	}
+
+	host, err := net.NewHost(&net.Config{
+		Ctx:        ctx,
+		DataDir:    conf.EnvConf.DataDir,
+		Port:       conf.Libp2pPort,
+		KeyFile:    conf.Libp2pKeyfile,
+		Bootnodes:  conf.EnvConf.Bootnodes,
+		ProtocolID: fmt.Sprintf("%s/%d", net.ProtocolID, chainID.Int64()),
+		ListenIP:   hostListenIP,
+		IsRelayer:  conf.IsRelayer,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if hostErr := host.Stop(); hostErr != nil {
+			err = multierror.Append(err, fmt.Errorf("error shutting down peer-to-peer services: %w", hostErr))
+		}
+	}()
+
+	swapBackend, err := backend.NewBackend(&backend.Config{
+		Ctx:                ctx,
+		MoneroClient:       conf.MoneroClient,
+		EthereumClient:     conf.EthereumClient,
+		Environment:        conf.EnvConf.Env,
+		SwapFactoryAddress: conf.EnvConf.SwapFactoryAddress,
+		SwapManager:        sm,
+		RecoveryDB:         sdb.RecoveryDB(),
+		Net:                host,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to make backend: %w", err)
+	}
+
+	log.Infof("created backend with monero endpoint %s and ethereum endpoint %s",
+		swapBackend.XMRClient().Endpoint(),
+		conf.EthereumClient.Endpoint(),
+	)
+
+	xmrTaker, err := xmrtaker.NewInstance(&xmrtaker.Config{
+		Backend:        swapBackend,
+		DataDir:        conf.EnvConf.DataDir,
+		NoTransferBack: conf.NoTransferBack,
+	})
+	if err != nil {
+		return err
+	}
+
+	xmrMaker, err := xmrmaker.NewInstance(&xmrmaker.Config{
+		Backend:  swapBackend,
+		DataDir:  conf.EnvConf.DataDir,
+		Database: sdb,
+		Network:  host,
+	})
+	if err != nil {
+		return err
+	}
+
+	// connect the maker/taker handlers to the p2p network host
+	host.SetHandlers(xmrMaker, xmrTaker)
+	if err = host.Start(); err != nil {
+		return err
+	}
+
+	rpcServer, err := rpc.NewServer(&rpc.Config{
+		Ctx:             ctx,
+		Address:         fmt.Sprintf("127.0.0.1:%d", conf.RPCPort),
+		Net:             host,
+		XMRTaker:        xmrTaker,
+		XMRMaker:        xmrMaker,
+		ProtocolBackend: swapBackend,
+	})
+
+	log.Infof("starting swapd with data-dir %s", conf.EnvConf.DataDir)
+	err = rpcServer.Start() // blocks until server is shutdown or context is cancelled
+	if err != nil && !errors.Is(err, http.ErrServerClosed) {
+		return err
+	}
+
+	// err can get set in defer blocks, so don't return nil below
+	return err
```

Comment:
I don't think this works, defer statements get evaluated after the return. see: https://go.dev/play/p/SbbM1nQj9ks

Reply:
I guess it makes sense, but I didn't realize that there is a subtle difference between named and unnamed return values in this regard.  From [the Go standard](https://go.dev/ref/spec):
> if the deferred function is a [function literal](https://go.dev/ref/spec#Function_literals) and the surrounding function has [named result parameters](https://go.dev/ref/spec#Function_types) that are in scope within the literal, the deferred function may access and modify the result parameters before they are returned.

I'll fix the code. Here's an alternative version of that example: https://go.dev/play/p/ajSRoozvuhK?v=goprev

Reply:
Done